abdiff/core/utils.py

```diff
@@ -124,9 +129,21 @@ def write_to_dataset(
     also be provided, because a schema cannot be extracted or inferred from the lazily
     evaluated generator.  All other types passed do not explicitly require a schema,
     as it can be inferred.
+
+    This function also dynamically sets option 'max_open_files' to align with current OS
+    level configurations open file limits, ensuring not to exceed, which can result in
+    "OSError: [Errno 24] Too many open files".
     """
-    written_files = []
+    # read open file limit and set max_open_files as a fraction
+    if not max_open_files:
+        soft_limit, hard_limit = resource.getrlimit(resource.RLIMIT_NOFILE)
+        max_open_files = max(int(soft_limit * 0.75), 50)
+        logger.debug(
+            f"OS open file limit of {soft_limit} detected, "
+            f"limiting max open files to {max_open_files}"
+        )
 
+    written_files = []
     ds.write_dataset(
```

Comment:
@ghukill When is a file closed? I figured that when the max number of rows have been written to a file, then `ds.write_dataset` would close it automatically. Is that not the case? Does it wait until it hits the `max_open_files` limit before closing any files -- as this description seems to suggest:

> If greater than 0 then this will limit the maximum number of files that can be left open. If an attempt is made to open too many files then the least recently used file will be closed. If this setting is set too low you may end up fragmenting your data into many small files

Reply:
The crux of the issue is partitioning.  Because we write the transformed files where each filename gets a partition, it _almost_ results in a dedicated partition (directory) for each input file, which could be thousands.

Normally, this many partitions is not ideal.  And we may even want to return to this and re-evaluate.  But it's helpful for joining, where we can limit by partition as a glob pattern.  We could also do this with individual files.

I _think_ the pyarrow dataset writer holds folder/files open for writing when it's partitioning.  If not partitioning, I think it would happily output 2k parquet files (roughly the length of the iterator for the large job).  But with partitions, I think it keeps those files open, as it may need to write to them.

I think we're close in a good architecture, but we may want to revisit this.  For example, even with the changes, it still consumes a lot of disk space.  I'm at 100+gb for transformed files only.  Granted they will get deleted as they are joined... we could theoretically performs joins right when the A and B containers are finished and avoid having those transformed files accumlate fully just to be deleted shortly thereafter.

Which is a long way of saying: **I think it's related to the partioning during writing, and I think we could/should revisit at some point.**